clientv3/balancer.go

```diff
@@ -26,32 +27,89 @@ import (
 // simpleBalancer does the bare minimum to expose multiple eps
 // to the grpc reconnection code path
 type simpleBalancer struct {
-	// eps are the client's endpoints stripped of any URL scheme
-	eps     []string
-	ch      chan []grpc.Address
-	numGets uint32
+	// addrs are the client's endpoints for grpc
+	addrs []grpc.Address
+	// notifyCh notifies grpc of the set of addresses for connecting
+	notifyCh chan []grpc.Address
+
+	// upc receives a message whenever up is posted
+	upc chan struct{}
+
+	// mu protects upEps, pinAddr, and connAddr
+	mu sync.Mutex
+	// upEps holds the current endpoints that have an active connection
+	upEps map[string]struct{}
+	// pinAddr is the currently pinned connected address
+	pinAddr string
+	// connAddr is the address proposed for the next connection
+	connAddr string
 }
 
-func newSimpleBalancer(eps []string) grpc.Balancer {
-	ch := make(chan []grpc.Address, 1)
+func newSimpleBalancer(eps []string) *simpleBalancer {
+	notifyCh := make(chan []grpc.Address, 1)
 	addrs := make([]grpc.Address, len(eps))
 	for i := range eps {
 		addrs[i].Addr = getHost(eps[i])
 	}
-	ch <- addrs
-	return &simpleBalancer{eps: eps, ch: ch}
+	notifyCh <- addrs
+	sb := &simpleBalancer{
+		addrs:    addrs,
+		notifyCh: notifyCh,
+		upc:      make(chan struct{}, 1),
+		upEps:    make(map[string]struct{}),
+	}
+	return sb
+}
+
+func (b *simpleBalancer) Start(target string) error { return nil }
+
+func (b *simpleBalancer) Up(addr grpc.Address) func(error) {
+	b.mu.Lock()
+	b.upEps[addr.Addr] = struct{}{}
+	if b.pinAddr == "" {
+		// pin first connected address
+		b.pinAddr = addr.Addr
+		b.connAddr = ""
+	}
+	b.mu.Unlock()
+	// notify client that a connection is up
+	select {
+	case b.upc <- struct{}{}:
```

Comment:
it seems like we only needs to send this once when we create the client? should we rename this to ready? and wrap it with sync.Once?